htmresearch/algorithms/synapse_learning.py

```diff
@@ -0,0 +1,241 @@
+#!/usr/bin/env python
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2017, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""Implementations of various synapse learning rules"""
+
+
+from abc import ABCMeta, abstractmethod
+
+import numpy as np
+
+
+
+class SynapseLearningRules(object):
+  __metaclass__ = ABCMeta
+
+  @abstractmethod
+  def learnOnExistingSegments(self, connections, learningSegments, activeAxons,
+                              growthCandidateAxons, **kwargs):
+    """
+    For each specified segment, reinforce / punish synapses according to whether
+    they're active, and maybe grow new synapses.
+
+    @param connections (any connections class)
+    Connections instance that will be modified
+
+    @param learningSegments (numpy array)
+    The segments to learn on
+
+    @param activeAxons (numpy array)
+    The active input to this Connections
+
+    @param growthCandidateAxons (numpy array)
+    The (presumably active) input bits that are available for growth
+
+    @param kwargs
+    Additional arguments for specific SynapseLearningRules implementations
+    """
+
+
+  @abstractmethod
+  def learnOnNewSegments(self, connections, learningCells, growthCandidateAxons,
+                         **kwargs):
+    """
+    Grow segments on the specified cells and grow synapses on each segment.
+
+    @param connections (any connections class)
+    Connections instance that will be modified
+
+    @param learningCells (numpy array)
+    The cells that should grow segments
+
+    @param growthCandidateAxons (numpy array)
+    The (presumably active) input bits that are available for growth
+
+    @param kwargs
+    Additional arguments for specific SynapseLearningRules implementations
+    """
+
+
+class ConnectToActiveAxons(SynapseLearningRules):
```

Comment:
Can we just call these "ActiveInputs" instead of axons? The term axon is bit weird in some contexts, such as encoders and spatial pooler.